File: day13/python/firewall.py
#!/usr/bin/python3

import sys
import itertools

def caught(depth, height):
    return (depth % ((height - 1) * 2)) == 0

if len(sys.argv) != 2:
    print('Provide input file as command line argument!')
else:
    file = open(sys.argv[1]).read().splitlines()

    firewall = {}
    for line in file:
        depth, height = map(int, line.split(': '))
        firewall[depth] = height

    severity = 0
    for layer in firewall:
        if caught(layer, firewall[layer]):
            severity += (layer * firewall[layer])

    print('Severity:', severity)

    for delay in itertools.count():
        stop = True
        for layer in firewall:
            if caught(layer + delay, firewall[layer]):
                stop = False
                break

        if stop:
            print('Delay:', delay)
            break

File: day13/rust/firewall.rs
use std::collections::HashMap;
use std::fs;

fn caught(depth: u32, height: u32) -> bool {
    (depth % ((height - 1) * 2)) == 0
}

fn main() {
    let content = fs::read_to_string("../input").expect("file not found");

    let mut firewall = HashMap::new();

    for line in content.lines() {
        let mut layer = line.split(": ");
        let depth = layer.next().unwrap().parse::<u32>().unwrap();
        let height = layer.next().unwrap().parse::<u32>().unwrap();
        firewall.insert(depth, height);
    }

    let mut severity = 0;
    for layer in &firewall {
        if caught(*layer.0, *layer.1) {
            severity += *layer.0 * *layer.1;
        }
    }

    println!("Severity: {}", severity);

    let mut delay = 0;
    loop {
        if firewall
            .iter()
            .all(|layer| !caught(*layer.0 + delay, *layer.1))
        {
            break;
        }
        delay += 1;
    }

    println!("{}", delay);
}
